README.md
# api-pagination-tester

Walk a cursor-paginated API modelled as an **in-memory mock** and prove what a client would end up
holding.

Pagination bugs are hard to see from inside a client. The page arrives, the records look fine, the
loop moves on — and the defect only shows up in the aggregate: one record appears twice because it
sat on a page boundary, another never appears at all, or the loop never ends because a cursor points
back at itself. This tool takes a scenario that declares both the **corpus** of record identities the
API is supposed to expose and the **pages** it serves, walks the state machine, and reports the
aggregate.

- **Repository:** [edilec/api-pagination-tester](https://github.com/edilec/api-pagination-tester)
- **Area:** API & Integration
- **License:** MIT

## Install and run

```
npx api-pagination-tester --input scenario.json
npx api-pagination-tester --input scenario.json --json > report.json
```

Node 22 or newer. **No runtime dependencies and no development dependencies** — Node built-ins only.

## What it finds

| | |
| --- | --- |
| **Boundary duplicates** | An identity served on two different pages. The client that appends every page to one list ends up holding it twice. |
| **Missing records** | A corpus identity served on no page of a **complete** walk. The usual cause is a strict comparison at the page boundary dropping one row. |
| **Repeated cursors** | A cursor handed out a second time. The walk terminates on a visited set and reports it; a client without one does not terminate at all. |
| **Empty pages vs last pages** | Zero records with a next cursor is an empty page; no next cursor is the end. A client that conflates them truncates the collection or refetches forever. |
| **Ambiguous terminators** | `nextCursor: ""`, which `if (next)` reads as the end and `if (next !== null)` reads as another page. |
| **Page-size drift** | A page over the size the API declares, or short before the last page. |
| **Unreachable pages** | A page the mock declares that nothing the API returns ever leads to, on a **complete** walk. A walk that stopped early never requested the rest of the chain either, so it claims nothing here. |
| **Unknown records** | An identity served that the corpus does not declare. |

## The scenario

```json
{
  "name": "orders",
  "style": "cursor",
  "pageLimit": 2,
  "corpus": ["ord-1", "ord-2", "ord-3", "ord-4", "ord-5"],
  "pages": [
    { "cursor": null,          "records": ["ord-1", "ord-2"], "nextCursor": "after:ord-2" },
    { "cursor": "after:ord-2", "records": ["ord-3", "ord-4"], "nextCursor": "after:ord-4" },
    { "cursor": "after:ord-4", "records": ["ord-5"],          "nextCursor": null }
  ]
}
```

`corpus` is the ground truth: without known identities a walk can only report what it saw, never
whether it saw everything. `cursor: null` is the initial request. `nextCursor` is **required** on
every page — an absent one is exactly the ambiguity this tool exists to find, so it is refused rather
than defaulted.

Three example scenarios ship with the package, one for each exit code:

```
node bin/api-pagination-tester.mjs --input examples/orders-clean.json    # exit 0, pass
node bin/api-pagination-tester.mjs --input examples/orders-broken.json   # exit 1, fail
node bin/api-pagination-tester.mjs --input examples/catalog-cycle.json   # exit 2, incomplete
```

The broken one holds a boundary duplicate, a missing record, an empty page in the middle of the
collection, and a page nothing leads to. The cycle one points a page back at itself: the walk
terminates, says so, and refuses to reach a completeness verdict it did not earn.

## Output

The human report:

```
orders-broken (cursor): 4 page(s) walked, 5 record(s) served, status fail.
walk: ended on last-page; completeness proven; corpus 5, distinct 4, duplicated 1, ...
page  cursor             records  next               note
0     (initial)          2        after:ord-2
1     after:ord-2        2        after:ord-3
2     after:ord-3        0        after:ord-4        empty
3     after:ord-4        1        (none)             last
ERROR   orders-broken.json/records/ord-4 record-missing The record "ord-4" is in the corpus ...
```

`--json` emits the machine-readable report instead. stdout carries the report and nothing else;
diagnostics go to stderr, so a non-empty stderr is normal and correct.

| Exit | Meaning |
| ---: | --- |
| `0` | The walk completed and the pagination contract held. |
| `1` | The walk completed and the contract failed. |
| `2` | Invalid usage or configuration (**stdout is empty**), or evidence that was missing, undecodable or bounded out (stdout carries a report with `status: "incomplete"`). |

## As a library

```js
import { testScenarioFile, walkPages } from 'api-pagination-tester'

const report = await testScenarioFile({ input: 'scenario.json' })

// Or walk your own in-process fake. `fetchPage` is a function, never a URL.
const result = await walkPages({
  fetchPage: async (cursor) => myFakeStore.page(cursor),
  limits: { maxPages: 100, maxRecords: 10000, maxMillis: 1000 },
  clock: () => performance.now(),
})
```

`testScenarioFile`, `testScenarioBytes`, `testScenarioText` and `testScenarioObject` all return the
same report. `walkPages` returns the raw walk — the pages in order, the cursors requested, and why it
stopped — with no verdict attached, which is what makes the state machine testable on its own.

## Limits and non-goals

**What this tool cannot conclude.**

- **It says nothing about a real HTTP API.** It walks the scenario you wrote, in memory. No socket is
  opened: `src/` imports no network module, and a library caller supplies an in-process function, not
  a URL. If the scenario does not match the service, the report describes the scenario — faithfully,
  and about the wrong thing.
- **It does not simulate concurrent writes.** Rows inserted or deleted mid-walk are the most common
  cause of real pagination duplicates and skips, and modelling them needs a timeline this tool does
  not have. A clean report here does not mean a clean walk under write load.
- **It cannot tell a missing record from a withheld one.** Authorisation, soft deletion, filtering
  and a genuine boundary bug all look identical from here: the identity was in the corpus and on no
  page.
- **One style only.** `cursor` is implemented. `offset`, `page-number`, `keyset` and `link-header`
  are recognised by name and reported as unsupported, which makes the run incomplete. An unsupported
  construct is never treated as satisfied.
- **A `pass` is not a stability claim.** It means this scenario's walk is self-consistent. It is not
  evidence that the ordering is stable, that cursors survive a deployment, or that the same walk
  holds tomorrow.
- **No latency, payload-size or rate-limit measurement.** Those need a transport, and there is none.
- **A bounded report, by design.** Every limit (bytes, declared pages, corpus size, pages walked,
  records, identity length, findings, milliseconds) is explicit and enforced. Exceeding one produces
  an `incomplete` report naming the limit — never a quietly shorter answer, and never a pass.

**What it will not do.**

- It never rewrites the scenario, and never writes anywhere but the file you named. `src/` imports
  no write API. `--out` is checked before the scenario is opened, and is refused when it is a
  symbolic link (`realpath` *resolves* a link, and resolving is the dangerous act — so it is
  refused on sight with `lstat`, whether or not its target exists yet), when a symlinked directory
  lies on the way to it, when it resolves outside `--out-root` (default: the working directory),
  when it is the scenario under any name (identity compared by inode, because a hard link is a
  second name for one file and resolves to a real path of its own), and when it is a directory or
  its directory does not exist. An existing regular file is refused as well without `--overwrite`,
  which is a separate question from safety and is asked after the guard.
- It never reports unknown as a pass. A walk that ended on anything but a real last page raises
  `completeness-unknown`, and no record is declared missing on the strength of a walk that never
  reached the end.
- It never guesses an encoding. Scenario bytes are decoded with `TextDecoder('utf-8', { fatal: true })`
  and refused if they are not UTF-8.
- It is deterministic. No `localeCompare`, no `Intl.Collator`, no `Date.now`, no `Math.random`; the
  clock is injected. Two runs over one scenario produce byte-identical stdout.
- It never quotes the scenario back. A scenario that does not parse is reported by position, line
  and column only. `JSON.parse` answers a short file by embedding the whole of it in the error
  message (`Unexpected token 'A', "AKIA..." is not valid JSON`) and a long one ten characters at a
  time, so a file that is only a credential would otherwise be reproduced by its own failure. That
  quoted window is removed before the message reaches a finding; sanitising alone does not remove it,
  because the window sits at the front of the message and truncation cuts from the back.
- It never echoes a control character. Every untrusted string reaching output — record ids and
  cursors included, not only excerpts — has C0, DEL, C1, U+2028/U+2029 and all twelve bidi controls replaced
  with a space. Sanitising happens on the way *out*; identity always uses the raw value, so the guard
  cannot manufacture a duplicate.

## Options

```
--input FILE            The scenario to run (required)
--root DIR              Report paths relative to this root, and refuse a scenario that
                        resolves outside it (default: the scenario's own directory)
--json                  Emit the machine-readable report on stdout
--out FILE              Also write the JSON report to FILE
--out-root DIR          Tree --out must resolve inside (default: the working directory)
--overwrite             Allow --out to replace an existing file
-h, --help              Show help

--max-scenario-bytes N  Bytes read from the scenario (default 4194304)
--max-scenario-pages N  Pages a scenario may declare (default 2000)
--max-corpus N          Record identities a corpus may declare (default 100000)
--max-pages N           Pages one walk may fetch (default 1000)
--max-records N         Records one walk may take in (default 100000)
--max-id-length N       Characters in a record id or cursor (default 200)
--max-findings N        Findings collected (default 1000)
--max-millis N          Milliseconds of walking, 0 for none (default 10000)
```

An unknown option is refused and a flag carrying a value may be given only once, so a typo cannot
quietly turn a real failure into a green run.

## Verification

```
npm run check      # lint, tests, the clean example, and a packaging dry run
```

`npm run lint` is `node --check` over every shipped file; `npm test` is `node --test`. There is
nothing to install.

The rule catalog, the limits, the report schema and the ordering rules are documented in
[docs/pagination-rules.md](./docs/pagination-rules.md).

## License

MIT. See [LICENSE](./LICENSE).


CHANGELOG.md
# Changelog

All notable changes to this project are recorded here. The format follows
[Keep a Changelog](https://keepachangelog.com/en/1.1.0/), and versions follow
[Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Rule ids are part of the public interface. Renaming one, or changing the severity of one, is a
breaking change and is recorded here.

## [Unreleased]

### Fixed

- A missing scenario linked to the named `--out` path, directly or through another
  symlink, could be turned into the report file while the run exited 2 as
  incomplete. The destination guard now resolves the scenario's final link
  chain before writing; an ordinary missing scenario with a separate report
  destination still produces an incomplete report.

- `--out` accepted a destination that destroyed a file the tool was never asked to touch. The check
  resolved the destination and compared the result with the scenario, which caught a symbolic link
  pointing AT the scenario and a hard link to it, and missed the case that actually loses data: a
  symbolic link pointing anywhere else. `realpath` resolved it, the resolved path was not the
  scenario, and the write went through the link. Measured with `--overwrite` — the flag whose whole
  purpose is to say "replace that file" — a 9-byte file outside the tree became a 1241-byte report
  at exit 0. A link whose target did not exist yet needed no `--overwrite` at all: the report was
  created outside the tree. A symlinked parent directory did the same thing one level up.
  `assertWritableDestination` now refuses all three before the scenario is opened, and the hard-link
  identity comparison lives inside the same guard. `test/destination.test.mjs` has one case per hole
  and one per allowed shape.

- A scenario that does not parse is no longer quoted back into the report. `JSON.parse` embeds the
  input in one of its two error messages — `Unexpected token 'A', "AKIA..." is not valid JSON` —
  so a scenario file short enough to be only a credential was reproduced in full by the
  `scenario-invalid` finding, on stdout, in both output modes. `parseFailureDetail` in `src/text.mjs`
  now removes the quoted window and keeps the position, line and column, which are the diagnostic
  half and carry no input. Sanitising and excerpting did not catch this: the quoted span is at the
  front of the message and both cut from the back.

### Added

- `--out-root`, declaring the tree `--out` may resolve inside. It defaults to the working
  directory and has no meaning without `--out`; `--overwrite` without `--out` is now a usage error
  too, rather than a flag with no effect.
- `assertWritableDestination` and `DestinationError`, exported for a caller writing its own
  destination logic.
- The pagination walker (`src/walk.mjs`): follows cursors from the initial request, remembers every
  cursor it has requested, and always terminates. Empty pages, last pages and the empty-string
  terminator are three separate states, not one.
- The in-memory mock API (`src/scenario.mjs`): a `Map` lookup in this process. No socket is opened
  and no port is bound, and `src/` imports no network module.
- Strict scenario validation: unknown keys refused rather than ignored, `nextCursor` required
  explicitly, exactly one initial page, no duplicate cursors, and a non-empty corpus of unique
  identities.
- The record-identity verdict: boundary duplicates, repeats inside one page, records the corpus does
  not declare, and — only for a walk that reached a real last page — records that are missing and
  declared pages nothing leads to. Both of those are inferences from having seen the whole chain, so
  a walk stopped by a loop, a dangling cursor or a bound claims neither.
- `completeness-unknown`: a walk that ended on a cycle, a dangling cursor or a bound proves nothing
  about completeness, so it says so and the run is incomplete rather than declaring unreached records
  missing.
- The rule catalog: 27 rules, each with its severity taken from one frozen table
  (`RULE_SEVERITY` in `src/rules.mjs`), asserted against `docs/pagination-rules.md` in both
  directions and pinned behaviourally by exit codes and literal inline counts.
- Eight explicit limits — scenario bytes, declared pages, corpus size, pages walked, records walked,
  identity length, findings and milliseconds — each enforced, each named in a finding when exceeded,
  and each tested from both sides of its bound. Exceeding one is an `incomplete` report, never a
  silent truncation.
- CLI `api-pagination-tester` with `--help`, `--json`, `--input`, `--root`, `--out`, `--overwrite`
  and a flag for every limit. Repeated value-carrying flags and unknown options are configuration
  errors, and a configuration error leaves stdout empty.
- `--out` refuses to be the scenario file, comparing `dev` and `ino` rather than resolved paths so
  that a hard link — which has no target for `realpath` to resolve — cannot launder the input into
  the output, and refuses to replace an existing file without `--overwrite`.
- Path confinement on real paths: both the root and the input are resolved before they are compared,
  so a symlink out of the root is refused and a root that is itself a symlink is not.
- Strict UTF-8 decoding of every scenario, through one function shared by every entry point.
- Output sanitising of C0, DEL, C1, U+2028/U+2029 and all twelve Unicode `Bidi_Control` characters
  (U+061C included) from every untrusted string, identifiers included. Sanitising is an output step;
  identity always uses the raw value.
- Determinism: ordering by UTF-16 code unit at every site, an injected clock, and byte-identical
  stdout over identical input.
- Three example scenarios, one for each exit code, and `docs/pagination-rules.md`.

No release has been published.


src/index.mjs
/**
 * api-pagination-tester
 *
 * Exercises a pagination state machine against an in-memory mock API built from
 * a scenario document, and reports what a client walking that API would end up
 * holding.
 *
 * Four properties are structural rather than incidental:
 *
 * 1. **No socket is ever opened.** The mock is a `Map` lookup in this process.
 *    A pagination contract is a property of what the responses say, and a
 *    transport would only add ways for the test to fail for reasons that are
 *    not about pagination. There is no network import in this package.
 * 2. **Records carry known identities.** The scenario declares the corpus the
 *    API is supposed to expose, which is what makes the two interesting
 *    verdicts possible at all: a record served on two pages is a boundary
 *    duplicate, and a corpus record served on no page is missing.
 * 3. **The walk always terminates.** Every requested cursor is remembered, so a
 *    cursor handed out twice ends the walk with a finding instead of spinning.
 * 4. **Unknown is never a pass.** A walk that stopped early -- a loop, a
 *    dangling cursor, a bound -- proves nothing about completeness, so it says
 *    so and the run is incomplete. Reporting unreached records as missing would
 *    be an assertion about evidence nobody obtained.
 */

import { readFile, realpath, stat } from 'node:fs/promises'
import { basename, dirname, relative, resolve, sep } from 'node:path'

import { analyzeWalk } from './analyze.mjs'
import { FindingSet, SECTIONS, createFinding, sortFindings } from './rules.mjs'
import { ScenarioError, createMockApi, validateScenario } from './scenario.mjs'
import { cursorLabel, decodeUtf8, excerpt, parseFailureDetail } from './text.mjs'

import { walkPages } from './walk.mjs'

export { DestinationError, assertWritableDestination } from './write-guard.mjs'

export const TOOL_ID = 'api-pagination-tester'
export const REPORT_SCHEMA_VERSION = '1'

/**
 * Bounds are part of the contract, not a safety net.
 *
 * A scenario is ordinary untrusted input, and a pagination walk is the one
 * shape of work that famously does not stop on its own. Every limit below is
 * explicit, overridable from the CLI, and named in a finding when it is hit.
 * Exceeding one produces an `incomplete` report -- never a quietly shorter
 * answer, and never a pass.
 *
 * `maxMillis` is the one limit that may be zero: the budget is spent once the
 * elapsed time reaches it, so zero is spent before the first fetch, which is
 * how the wiring between the flag and the clock is proven rather than assumed.
 */
export const DEFAULT_LIMITS = Object.freeze({
  maxCorpus: 100000,
  maxFindings: 1000,
  maxIdLength: 200,
  maxMillis: 10000,
  maxPages: 1000,
  maxRecords: 100000,
  maxScenarioBytes: 4194304,
  maxScenarioPages: 2000,
})

const ZERO_ALLOWED_LIMITS = Object.freeze(['maxMillis'])

const FILE_OPTIONS = Object.freeze(['clock', 'input', 'limits', 'root'])
const MEMORY_OPTIONS = Object.freeze(['clock', 'file', 'limits'])

function isRecord(value) {
  return Boolean(value) && typeof value === 'object' && !Array.isArray(value)
}

function rejectUnknownOptions(options, allowed) {
  if (!isRecord(options)) throw new TypeError('Options must be an object')
  for (const key of Object.keys(options)) {
    if (!allowed.includes(key)) throw new TypeError(`Unknown option "${key}"`)
  }
}

/**
 * Only an absent `limits` means "use the defaults".
 *
 * `null` is a value the caller computed and lost, not an omission, and
 * accepting it as `{}` is the same silent ignore this tool refuses everywhere
 * else: an unknown limit name and a fractional limit are both errors, so a
 * limits object that turned out to be null cannot be the one thing waved
 * through. A misspelled limit must not quietly leave the default in place.
 */
export function validateLimits(overrides = {}) {
  if (!isRecord(overrides)) throw new TypeError('Limits must be an object')
  const limits = { ...DEFAULT_LIMITS }
  for (const [name, value] of Object.entries(overrides)) {
    if (!Object.hasOwn(DEFAULT_LIMITS, name)) throw new TypeError(`Unknown limit "${name}"`)
    const floor = ZERO_ALLOWED_LIMITS.includes(name) ? 0 : 1
    if (!Number.isInteger(value) || value < floor) {
      throw new TypeError(`Limit "${name}" must be an integer of at least ${floor}`)
    }
    limits[name] = value
  }
  return Object.freeze(limits)
}

function validateClock(clock) {
  if (clock === undefined) return () => performance.now()
  if (typeof clock !== 'function') throw new TypeError('Clock must be a function returning milliseconds')
  return clock
}

/**
 * Containment, decided on real paths.
 *
 * Refusing `../` and absolute strings is not confinement: a symbolic link
 * planted inside the declared root resolves out of the tree without ever
 * spelling a traversal. Both sides of this comparison have been through
 * `realpath` before they arrive -- comparing a real root against a path that
 * was not resolved is the over-correction, and it refuses scenarios that
 * genuinely are inside a root reached through a symlink. A false refusal is a
 * bug too.
 */
export function isInside(root, candidate) {
  if (candidate === root) return true
  return candidate.startsWith(root.endsWith(sep) ? root : `${root}${sep}`)
}

const EMPTY_VERDICT = Object.freeze({
  duplicated: [],
  missing: [],
  repeatedInPage: [],
  unknown: [],
  unreachablePages: [],
  distinct: 0,
  observedCorpus: 0,
})

const EMPTY_WALK = Object.freeze({
  pages: [],
  requested: [],
  served: 0,
  terminatedBy: null,
  repeatedCursor: null,
  danglingCursor: null,
  ambiguousTerminator: false,
  ambiguousFollows: false,
  probedEmptyCursor: false,
  complete: false,
})

/** Every id and cursor in the report is sanitised and bounded on the way out. */
const label = (value) => excerpt(value, 120)

function buildReport({ findings, scenario, result, verdict, bytes }) {
  const rows = sortFindings(findings.rows).map(createFinding)
  const errors = rows.filter((finding) => finding.severity === 'error').length
  const warnings = rows.filter((finding) => finding.severity === 'warning').length
  const status = findings.incomplete ? 'incomplete' : errors > 0 ? 'fail' : 'pass'

  return {
    schemaVersion: REPORT_SCHEMA_VERSION,
    tool: TOOL_ID,
    status,
    summary: {
      checked: result.pages.length,
      errors,
      warnings,
      info: rows.length - errors - warnings,
      pages: result.pages.length,
      records: result.served,
      distinctRecords: verdict.distinct,
      observedCorpus: verdict.observedCorpus,
      corpus: scenario === null ? 0 : scenario.corpus.length,
      duplicated: verdict.duplicated.length,
      repeatedInPage: verdict.repeatedInPage.length,
      missing: verdict.missing.length,
      unknown: verdict.unknown.length,
      unreachablePages: verdict.unreachablePages.length,
      bytes,
    },
    walk: {
      scenario: scenario === null ? null : label(scenario.name),
      style: scenario === null ? null : label(scenario.style),
      pageLimit: scenario === null ? null : scenario.pageLimit,
      terminatedBy: result.terminatedBy,
      complete: result.complete,
      pages: result.pages.map((page) => ({
        index: page.index,
        cursor: page.cursor === null ? null : label(page.cursor),
        records: page.records.length,
        nextCursor: page.nextCursor === null ? null : label(page.nextCursor),
        empty: page.empty,
        terminal: page.terminal,
      })),
      records: {
        duplicated: verdict.duplicated.map(label),
        repeatedInPage: verdict.repeatedInPage.map(label),
        missing: verdict.missing.map(label),
        unknown: verdict.unknown.map(label),
      },
      unreachablePages: verdict.unreachablePages.map(label),
    },
    findings: rows,
  }
}

/**
 * A report for a scenario this tool refused to run.
 *
 * The single place where a scenario-level refusal marks the run incomplete.
 * Every such refusal arrives here, so none of them can acquire a different idea
 * of what "we obtained no evidence" means, and deleting this one assignment
 * changes the exit code of every one of them.
 */
function refusedReport(error, file, bytes) {
  const findings = new FindingSet({ file, maxFindings: DEFAULT_LIMITS.maxFindings })
  findings.add(error.ruleId, error.message, {
    section: SECTIONS.scenario,
    pointer: error.pointer,
    suggestion: error.suggestion,
    evidence: error.evidence,
  })
  findings.incomplete = true
  return buildReport({ findings, scenario: null, result: EMPTY_WALK, verdict: EMPTY_VERDICT, bytes })
}

/**
 * Run a scenario that is already a JavaScript object.
 *
 * The innermost entry point: every other one decodes, parses and then arrives
 * here, so the validation, the walk and the verdict cannot differ between them.
 */
export async function testScenarioObject(document, options = {}) {
  rejectUnknownOptions(options, MEMORY_OPTIONS)
  const file = options.file === undefined ? 'scenario.json' : options.file
  if (typeof file !== 'string' || file.trim() === '') throw new TypeError('File label must be a non-empty string')
  const limits = validateLimits(options.limits)
  const clock = validateClock(options.clock)
  return runScenario(document, { file: excerpt(file, 200), limits, clock, bytes: 0 })
}

async function runScenario(document, { file, limits, clock, bytes }) {
  let scenario
  try {
    scenario = validateScenario(document, limits)
  } catch (error) {
    if (error instanceof ScenarioError) return refusedReport(error, file, bytes)
    throw error
  }

  const findings = new FindingSet({ file, maxFindings: limits.maxFindings })
  const api = createMockApi(scenario)
  const result = await walkPages({ fetchPage: (cursor) => api.fetchPage(cursor), limits, clock })
  const verdict = analyzeWalk({ scenario, result, findings })
  return buildReport({ findings, scenario, result, verdict, bytes })
}

/**
 * Run a scenario from bytes.
 *
 * Decoding is strict and happens here, once, for every caller that has bytes:
 * a scenario that is not UTF-8 is refused by the decoder rather than inferred
 * from decoded text afterwards. There is no second, looser path -- the CLI, the
 * file entry point and this one all decode through `decodeUtf8`.
 */
export async function testScenarioBytes(bytes, options = {}) {
  if (!(bytes instanceof Uint8Array)) throw new TypeError('Bytes must be a Uint8Array')
  rejectUnknownOptions(options, MEMORY_OPTIONS)
  const file = options.file === undefined ? 'scenario.json' : options.file
  if (typeof file !== 'string' || file.trim() === '') throw new TypeError('File label must be a non-empty string')
  const limits = validateLimits(options.limits)
  const clock = validateClock(options.clock)
  const fileLabel = excerpt(file, 200)

  if (bytes.byteLength > limits.maxScenarioBytes) {
    return refusedReport(
      new ScenarioError(
        'scenario-too-large',
        `The scenario is ${bytes.byteLength} byte(s), past the maxScenarioBytes limit of ` +
          `${limits.maxScenarioBytes}. It was not read, so nothing was exercised.`,
        { suggestion: 'Raise --max-scenario-bytes, or split the scenario.' },
      ),
      fileLabel,
      bytes.byteLength,
    )
  }

  let text
  try {
    text = decodeUtf8(bytes)
  } catch {
    return refusedReport(
      new ScenarioError(
        'scenario-not-utf8',
        'The scenario is not valid UTF-8, so it was refused by the decoder rather than guessed at. Nothing in it ' +
          'was read.',
        { suggestion: 'Re-save the scenario as UTF-8. A scenario read through the wrong encoding describes an API nobody has.' },
      ),
      fileLabel,
      bytes.byteLength,
    )
  }

  let document
  try {
    document = JSON.parse(text)
  } catch (error) {
    return refusedReport(
      new ScenarioError('scenario-invalid', `The scenario is not valid JSON: ${excerpt(parseFailureDetail(error), 160)}`, {
        suggestion: 'Check the scenario against docs/pagination-rules.md.',
      }),
      fileLabel,
      bytes.byteLength,
    )
  }

  return runScenario(document, { file: fileLabel, limits, clock, bytes: bytes.byteLength })
}

/** Run a scenario from text that does not live on disk. No filesystem access. */
export async function testScenarioText(text, options = {}) {
  if (typeof text !== 'string') throw new TypeError('Scenario text must be a string')
  return testScenarioBytes(new TextEncoder().encode(text), options)
}

/**
 * Run a scenario from a file.
 *
 * The file is opened read-only and never written to: this module imports no
 * write API at all. The reported `location.file` is relative to the declared
 * root, so an absolute host path never reaches the report.
 */
export async function testScenarioFile(options = {}) {
  rejectUnknownOptions(options, FILE_OPTIONS)
  if (typeof options.input !== 'string' || options.input.trim() === '') {
    throw new TypeError('An input path is required')
  }
  if (options.root !== undefined && (typeof options.root !== 'string' || options.root.trim() === '')) {
    throw new TypeError('Root must be a non-empty string')
  }
  const limits = validateLimits(options.limits)
  const clock = validateClock(options.clock)

  /**
   * The root is resolved before the input.
   *
   * A root that cannot be read is a configuration error whatever the input
   * turns out to be, and a configuration error is a run that never had a
   * subject. Resolving the input first would let a missing input turn a
   * misspelled root into an ordinary incomplete report, which is the quieter
   * and more misleading of the two answers.
   */
  let rootReal = null
  if (options.root !== undefined) {
    try {
      rootReal = await realpath(resolve(options.root))
    } catch (error) {
      throw new TypeError(`Root could not be read: ${error.code ?? 'unknown error'}`)
    }
  }

  let inputReal
  try {
    inputReal = await realpath(resolve(options.input))
  } catch (error) {
    return refusedReport(
      new ScenarioError(
        'scenario-unreadable',
        `The scenario could not be read: the path could not be resolved (${error.code ?? 'unknown error'}).`,
        { suggestion: 'Check the path and its permissions.' },
      ),
      excerpt(basename(options.input), 200),
      0,
    )
  }
  if (rootReal === null) rootReal = dirname(inputReal)

  const fileLabel = excerpt(relative(rootReal, inputReal) || basename(inputReal), 200)

  if (!isInside(rootReal, inputReal)) {
    return refusedReport(
      new ScenarioError(
        'scenario-unreadable',
        'The scenario resolves outside the declared root, so it was refused unread. A symbolic link inside a root ' +
          'is still a path out of it.',
        { suggestion: 'Point --root at the tree the scenario really lives in, or move the file inside it.' },
      ),
      excerpt(basename(inputReal), 200),
      0,
    )
  }

  const info = await stat(inputReal).catch(() => null)
  if (info === null || !info.isFile()) {
    return refusedReport(
      new ScenarioError('scenario-unreadable', 'The scenario could not be read: the path is not a regular file.', {
        suggestion: 'Point --input at a scenario file.',
      }),
      fileLabel,
      0,
    )
  }
  if (info.size > limits.maxScenarioBytes) {
    return refusedReport(
      new ScenarioError(
        'scenario-too-large',
        `The scenario is ${info.size} byte(s), past the maxScenarioBytes limit of ${limits.maxScenarioBytes}. It ` +
          'was not read, so nothing was exercised.',
        { suggestion: 'Raise --max-scenario-bytes, or split the scenario.' },
      ),
      fileLabel,
      info.size,
    )
  }

  let bytes
  try {
    bytes = await readFile(inputReal)
  } catch (error) {
    return refusedReport(
      new ScenarioError('scenario-unreadable', `The scenario could not be read: ${error.code ?? error.message}.`, {
        suggestion: 'Check the path and its permissions.',
      }),
      fileLabel,
      0,
    )
  }

  return testScenarioBytes(new Uint8Array(bytes), { file: fileLabel, limits: { ...limits }, clock })
}

const SEVERITY_WIDTH = 7
const CURSOR_WIDTH = 18

/** The human summary. Every id, cursor and message in it has been sanitised. */
export function formatReport(report) {
  const { summary, walk } = report
  const lines = [
    `${walk.scenario ?? 'scenario'} (${walk.style ?? 'unknown style'}): ${summary.pages} page(s) walked, ` +
      `${summary.records} record(s) served, status ${report.status}.`,
    `walk: ended on ${walk.terminatedBy ?? 'nothing (the scenario was refused)'}; completeness ` +
      `${walk.complete ? 'proven' : 'not proven'}; corpus ${summary.corpus}, distinct ${summary.distinctRecords}, ` +
      `duplicated ${summary.duplicated}, repeated-in-page ${summary.repeatedInPage}, missing ${summary.missing}, ` +
      `unknown ${summary.unknown}, unreachable pages ${summary.unreachablePages}.`,
  ]

  if (walk.pages.length > 0) {
    lines.push(`page  ${'cursor'.padEnd(CURSOR_WIDTH)} records  ${'next'.padEnd(CURSOR_WIDTH)} note`)
    for (const page of walk.pages) {
      const note = page.terminal ? (page.empty ? 'last, empty' : 'last') : page.empty ? 'empty' : ''
      const next = page.nextCursor === null ? '(none)' : cursorLabel(page.nextCursor)
      lines.push(
        `${String(page.index).padEnd(5)} ${cursorLabel(page.cursor).padEnd(CURSOR_WIDTH)} ` +
          `${String(page.records).padEnd(8)} ${next.padEnd(CURSOR_WIDTH)} ${note}`.trimEnd(),
      )
    }
  }

  for (const finding of report.findings) {
    const place = `${finding.location.file}${finding.location.pointer}`
    const quoted = finding.evidence === undefined ? '' : ` -- ${finding.evidence}`
    lines.push(
      `${finding.severity.toUpperCase().padEnd(SEVERITY_WIDTH)} ${place} ${finding.ruleId} ${finding.message}${quoted}`,
    )
  }
  return `${lines.join('\n')}\n`
}

export { analyzeWalk, indexRecords } from './analyze.mjs'
export { FindingSet, RULE_SEVERITY, SECTIONS, SEVERITY_VALUES, createFinding, sortFindings } from './rules.mjs'
export {
  KNOWN_UNSUPPORTED_STYLES,
  SUPPORTED_STYLES,
  ScenarioError,
  createMockApi,
  validateScenario,
} from './scenario.mjs'
export {
  EXCERPT_LIMIT,
  FORGEABLE_RANGES,
  byCodeUnit,
  cursorLabel,
  decodeUtf8,
  excerpt,
  isForgeable,
  parseFailureDetail,
  sanitize,
} from './text.mjs'
export { TERMINATIONS, walkPages } from './walk.mjs'
